translated/tech/20190106-Go-GraphQL-Beginners-Tutorial-Part1.md

```diff
@@ -0,0 +1,398 @@
+                        // we want to be able to specify the ID of the
+                        // tutorial we want to retrieve
+                        Args: graphql.FieldConfigArgument{
+                                "id": &graphql.ArgumentConfig{
+                                        Type: graphql.Int,
+                                },
+                        },
+                        Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+                                // take in the ID argument
+                                id, ok := p.Args["id"].(int)
+                                if ok {
+                                        // Parse our tutorial array for the matching id
+                                        for _, tutorial := range tutorials {
+                                                if int(tutorial.ID) == id {
+                                                        // return our tutorial
+                                                        return tutorial, nil
+                                                }
+                                        }
+                                }
+                                return nil, nil
+                        },
+                },
+                // this is our `list` endpoint which will return all
+                // tutorials available
+                "list": &graphql.Field{
+                        Type:        graphql.NewList(tutorialType),
+                        Description: "Get Tutorial List",
+                        Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+                                return tutorials, nil
+                        },
+                },
+        }
+```
+
+所以我们创建了类型并更新了GraphQL模式(Schema)，我们做的还算不错。
+
+## 测试它是否能够工作
+
+让我们尝试使用新的GraphQL服务器，并使用我们提交的查询。让我们通过更改`main()`函数中的`query`来尝试`list`模式。
+
+```go
+// Query
+query := `
+    {
+        list {
+            id
+            title
+            comments {
+                body
+            }
+            author {
+                Name
+                Tutorials
+            }
+        }
+    }
+`
+```
+
+让我们来分析下，在我们的查询中我们有一个特殊的`root`对象。在这里，我们说我们想要的那个对象的`list`字段，在按`list`返回的列表中，我们希望能看到`id`，`title`，`comments` 和`author`。
+
+而当我们运行它时，我们将会看到如下输出：
+
+```bash
+$ go run ./...
+{"data":{"list":[{"author":{"Name":"Elliot Forbes","Tutorials":[1]},"comments":[{"body":"First Comment"}],"id":1,"title":"Go GraphQL Tutorial"}]}}
+```
+
+正如我们所看到的，我们的查询以JSON形式返回了所有教程，看起来和初始的查询非常相似。
+
+现在让我们通过`tutorial`模式来执行一个查询:
+
+```go
+query := `
+    {
+        tutorial(id:1) {
+            title
+            author {
+                Name
+                Tutorials
+            }
+        }
+    }
+`
+```
+
+再一次，当我们运行它，我们会看到它成功地检索到了内从中`ID=1`的单独的教程。
+
+```bash
+$ go run ./...
+{"data":{"tutorial":{"author":{"Name":"Elliot Forbes","Tutorials":[1]},"title":"Go GraphQL Tutorial"}}}
+```
+
+完美，看起来我们的`list`和`tutorial`模式能够正常工作。
+
```

Comment:
缺了内容：
Challenge - Try updating the list of tutorials within our populate() function so that it returns more tutorials. Once we’ve done this, play around with the queries and try to become more familiar with them.